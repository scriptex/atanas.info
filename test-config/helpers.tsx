import { act } from 'react-dom/test-utils';
import type { FC } from 'react';
import { render, RenderResult, waitFor } from '@testing-library/react';

import * as shared from '@scripts/shared';

type ComponentType = FC<Readonly<React.ComponentProps<any>>>;

export const test = async (Component: ComponentType): Promise<RenderResult> => {
	let result!: RenderResult;

	await act(async () => {
		result = await waitFor(() => render(<Component />));
	});

	return result;
};

export const snapshotTest = (Component: ComponentType, buttonSelector?: string): void => {
	const { name } = Component;

	jest.spyOn(shared, 'formatDate');

	(shared.formatDate as jest.Mock).mockImplementation(() => 'Mock date');

	afterAll(() => {
		jest.clearAllMocks();
	});

	describe(`${name} component`, () => {
		it(`Should render the ${name} component`, async () => {
			const { asFragment, container } = await test(Component);

			expect(asFragment()).toMatchSnapshot();

			if (buttonSelector) {
				await act(async () => {
					await container.querySelectorAll<HTMLElement>(buttonSelector).forEach(button => {
						if (!button.hasAttribute('download')) {
							button.click();

							expect(asFragment()).toMatchSnapshot();
						}
					});
				});
			}
		});

		it(`Should unmount the ${name} component`, async () => {
			const { unmount, asFragment } = await test(Component);

			unmount();

			expect(asFragment()).toMatchSnapshot();
		});
	});
};

export const mockAudioContext = (): jest.Mock => {
	const mockConnect = jest.fn();

	const mockgetByteFrequencyData = jest.fn();

	const mockcreateMediaElementSource = jest.fn(() => ({
		connect: mockConnect
	}));

	const mockcreateAnalyser = jest.fn(() => ({
		connect: mockConnect,
		frequencyBinCount: [0, 1, 2],
		getByteFrequencyData: mockgetByteFrequencyData
	}));

	const mockcreateOscillator = jest.fn(() => ({
		channelCount: 2
	}));

	const mockChannelSplitterConnect = jest.fn(n => n);

	const mockcreateChannelSplitter = jest.fn(() => ({
		connect: mockChannelSplitterConnect
	}));

	const mockedAudioContext = jest.fn(() => ({
		createAnalyser: mockcreateAnalyser,
		createOscillator: mockcreateOscillator,
		createChannelSplitter: mockcreateChannelSplitter,
		createMediaElementSource: mockcreateMediaElementSource
	}));

	return mockedAudioContext;
};

export const mockFetch = (value: Record<string, string | boolean>): void => {
	global.fetch = jest.fn().mockImplementation(() =>
		Promise.resolve({
			json: () => Promise.resolve(value)
		})
	);
};
