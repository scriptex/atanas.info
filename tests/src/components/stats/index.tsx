import { StatsEntry, StatsError } from '@components';
import { snapshotTest } from '@test-config/helpers';

snapshotTest(() => <StatsError network="Github" />, undefined, 'StatsError');

snapshotTest(
	() => <StatsEntry data={[{ index: 0, title: 'Test title', value: 'Test value' }]} title="Test title" />,
	undefined,
	'StatsEntry'
);
