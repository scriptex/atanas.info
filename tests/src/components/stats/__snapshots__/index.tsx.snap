// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`StatsEntry component Should render the StatsEntry component 1`] = `
<DocumentFragment>
  <div
    class="c-section__entry c-section__entry--no-background"
  >
    <div
      class="o-shell"
    >
      <h3>
        Test title
      </h3>
      <ul
        class="c-section__list"
      >
        <li>
          <span>
            Test title:
          </span>
          <strong>
            Test value
          </strong>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`StatsEntry component Should unmount the StatsEntry component 1`] = `<DocumentFragment />`;

exports[`StatsError component Should render the StatsError component 1`] = `
<DocumentFragment>
  <div
    class="c-section__entry c-section__entry--no-background"
  >
    <div
      class="o-shell"
    >
      <h3>
        Github profile statistics
      </h3>
      <p>
        Failed fetching data from Github. Please check again in 8 hours.
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`StatsError component Should unmount the StatsError component 1`] = `<DocumentFragment />`;
